### PR TITLE
Add new steps, traits and helper methods

### PR DESCRIPTION
**Pull request by bakoe**
See diff of `CHANGELOG.md` for an overview of the added/changed functionality.

### CODE REVIEW

File: src/ConditionSteps.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+use \Behat\Behat\Tester\Exception\PendingException;
+
+trait ConditionSteps {
+
+    /**
+     * Check if the current time is within the specified time. Otherwise, throw a PendingException and thus skip the test case.
+     * Example: When the current time is between "06:00" and "20:00", otherwise skip the test case
+     *
+     * @When /^the current time is between "(?P<fromTime>[^"]+)" and "(?P<toTime>[^"]+)", otherwise skip the test case/
+     * @When /^die aktuelle Uhrzeit liegt zwischen "(?P<fromTime>[^"]+)" und "(?P<toTime>[^"]+)", sonst breche das Testzsenario ab/
```
Comment (OskarStark) [inline src/ConditionSteps.php]: Typo in Testszenario

File: src/WebsiteInteractionSteps.php [hunk 2]
```diff
@@ -21,19 +21,55 @@ public function getDefaultTimeoutWebsiteInteraction()
...
-     * Scrolls the element matching the given selector into view
+     * Returns the default widths in pixels used when resizing the viewport.
+     * You may override the default responsive widths in classes using this trait by using:
+     *   protected $defaultResponsiveViewportWidths = array(
+     *        "Desktop" => 1365,
+     *        "Tablet"  => 1023,
+     *        "Phone"   => 640
+     *   );
+     *
+     * œparam string deviceType one of "Desktop", "Tablet" or "Phone"
```
Comment (OskarStark) [inline src/WebsiteInteractionSteps.php]: @param looks weird here

Comment (bakoe) [reply src/WebsiteInteractionSteps.php]: Oops! Thanks for noticing!

File: CHANGELOG.md [hunk 3]
```diff
@@ -7,7 +7,16 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## Unreleased
 ### Added
 - Add "openUrl" and "openUrlWithinSpecifiedTime" steps. Both open a url and compares, within a specified time, the current url with the expected one
+- Add "ConditionSteps" trait for conditional steps that mark tests as "pending" when specific conditions are not met
 - Add "actualTimeIsInSpecifiedTime" step that allows checking whether the current time is in the expected range
+- Add "resizeWindowToResponsiveViewport" step that allows resizing the driver's browser windows for a given responsive device type
```
Comment (OskarStark) [inline CHANGELOG.md]:
```suggestion
- Add `resizeWindowToResponsiveViewport` step that allows resizing the driver's browser windows for a given responsive device type
```

but this should be done across the whole file in a separate PR

File: features/bootstrap/FeatureContext.php [hunk 4]
```diff
@@ -4,6 +4,7 @@ class FeatureContext extends \Behat\MinkExtension\Context\MinkContext {
     use \Exozet\Behat\Utils\Base\JsonApiSteps;
     use \Exozet\Behat\Utils\Base\WebsiteInteractionSteps;
     use \Exozet\Behat\Utils\Base\SpinnedMinkSteps;
+    use \Exozet\Behat\Utils\Base\ConditionSteps;
 
     protected $defaultTimeout = 3;  // Set the default timeout to 3 seconds
 }
```
Comment (OskarStark) [inline features/bootstrap/FeatureContext.php]: please add a newline to get the red pictogram removed

File: src/ConditionSteps.php [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        if($this->checkTime($currentTime,$fromDate, $toDate) == false){
+            throw new PendingException("The current time (" . $currentTime . ") is outside of the specified range. Specified range: from " . $fromDate . " to " .$toDate);
+        }
+    }
+
+    /**
+     * Check if the current time is within the specified time and return true or false
+     * Example: currentTime: 22:34 - fromeDate: 20:00 - toDate: 06:00 -> Result: false
+     */
+    private function checkTime($currentTime, $fromDate, $toDate){
```
Comment (OskarStark) [inline src/ConditionSteps.php]: this function does not check if `fromDate` and `toDate` are equal. and `currentTime` is equal, too:

For example: `06:00`, `06:00`, `06:00` should be `true`!

File: src/ConditionSteps.php [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        $toDate = date("H:i", strtotime($toTime));
+
+        if($this->checkTime($currentTime,$fromDate, $toDate) == false){
+            throw new PendingException("The current time (" . $currentTime . ") is outside of the specified range. Specified range: from " . $fromDate . " to " .$toDate);
+        }
+    }
+
+    /**
+     * Check if the current time is within the specified time and return true or false
+     * Example: currentTime: 22:34 - fromeDate: 20:00 - toDate: 06:00 -> Result: false
```
Comment (OskarStark) [inline src/ConditionSteps.php]: shouldn't this example return in `true`? 🤔

File: src/HelpUtils.php [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    /**
+     * Returns the default timeout in milliseconds used by JavaScript steps that involve a timeout.
+     * You may override the default timeout in classes using this trait by using:
+     *   protected $defaultJavaScriptTimeout = 10;  // Set the default timeout to 10 seconds
+     *
+     * @return int the default timeout in milliseconds
+     * @throws ExpectationException
+     */
+    public function getDefaultJavascriptTimeInMilliseconds()
+    {
```
Comment (OskarStark) [inline src/HelpUtils.php]: here the opening bracket is on a new line....

File: src/HelpUtils.php [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        $timeoutInSeconds = isset($this->defaultJavascriptTimeout) ? $this->defaultJavascriptTimeout : 5000;
+        return $timeoutInSeconds * 1000;
+    }
+
+    public function getDefaultJavascriptTime() {
```
Comment (OskarStark) [inline src/HelpUtils.php]: .. here the opening bracket is on the same line...

File: src/WebsiteInteractionSteps.php [hunk 9]
```diff
@@ -3,8 +3,8 @@
 namespace Exozet\Behat\Utils\Base;
 
 use \Behat\Gherkin\Node\PyStringNode;
-use Behat\Mink\Exception\ElementNotFoundException;
-use Behat\Mink\Exception\ExpectationException;
+use \Behat\Mink\Exception\ElementNotFoundException;
+use \Behat\Mink\Exception\ExpectationException;
```
Comment (OskarStark) [inline src/WebsiteInteractionSteps.php]: These leading backslahshes should be removed

File: src/WebsiteInteractionSteps.php [hunk 10]
```diff
@@ -21,19 +21,55 @@ public function getDefaultTimeoutWebsiteInteraction()
...
+     *
      * @Given /^I scroll to "(?P<selector>[^"]+)"$/
      * @Given /^ich scrolle zu "(?P<selector>[^"]+)"$/
      */
-    public function scrollIntoView($selector)
+    public function scrollIntoView($selector, $alignToTop = true)
     {
         $this->getSession()->executeScript(
-            'document.querySelectorAll(' . json_encode($selector) . ')[0].scrollIntoView()'
+            'document.querySelectorAll(' . json_encode($selector) . ')[0].scrollIntoView(' . ($alignToTop) ? "true" : "false" . ')'
```
Comment (OskarStark) [inline src/WebsiteInteractionSteps.php]: I would prefer creating this "string" via `sprintf()`, but its a personal opinion

File: src/ConditionSteps.php [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+use \Behat\Behat\Tester\Exception\PendingException;
+
+trait ConditionSteps {
+
+    /**
+     * Check if the current time is within the specified time. Otherwise, throw a PendingException and thus skip the test case.
+     * Example: When the current time is between "06:00" and "20:00", otherwise skip the test case
+     *
+     * @When /^the current time is between "(?P<fromTime>[^"]+)" and "(?P<toTime>[^"]+)", otherwise skip the test case/
```
Comment (OskarStark) [inline src/ConditionSteps.php]:
```suggestion
     * @When /^the current time is between "(?P<fromTime>[^"]+)" and "(?P<toTime>[^"]+)", otherwise skip the test case$/
```

File: src/ConditionSteps.php [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+     * @When /^die aktuelle Uhrzeit liegt zwischen "(?P<fromTime>[^"]+)" und "(?P<toTime>[^"]+)", sonst breche das Testszenario ab/
```
Comment (OskarStark) [inline src/ConditionSteps.php]:
```suggestion
     * @When /^die aktuelle Uhrzeit liegt zwischen "(?P<fromTime>[^"]+)" und "(?P<toTime>[^"]+)", sonst breche das Testszenario ab$/
```

File: src/WebsiteInteractionSteps.php [hunk 13]
```diff
@@ -377,44 +413,48 @@ public function openUrlWithinSpecifiedTime($page, $seconds)
...
-            if(($currentTime >= $fromDate) || ($currentTime <= $toDate)) {
-                return true;
-            }else {
-                return false;
-            }
+    /**
+     * Hover over a given element, i. e. toggles the corresponding mouse-over event
+     * Example: When I hover over the element "button.send"
+     *
+     * @When /^I hover over the element "(?P<selector>[^"]+)"$/
```
Comment (OskarStark) [inline src/WebsiteInteractionSteps.php]:
```suggestion
     * @When /^I hover the element "(?P<selector>[^"]+)"$/
```

??

File: src/WebsiteInteractionSteps.php [hunk 14]
```diff
@@ -377,44 +413,48 @@ public function openUrlWithinSpecifiedTime($page, $seconds)
...
+     * @When /^ich über das Element "(?P<selector>[^"]+)" hovere$/
+     * @throws ElementNotFoundException
+     */
+    public function mouseOver($selector)
+    {
+        $domElement = $this->getSession()->getPage()->find("css", $selector);
```
Comment (OskarStark) [inline src/WebsiteInteractionSteps.php]: You should use `'`(single quotes wherever you can

> PHP won't use additional processing to interpret what is inside the single quote. when you use double quotes PHP has to parse to check if there is any variables in there. A single-quoted string does not have variables within it interpreted. A double-quoted string does.

File: CHANGELOG.md (hunk 3 above)
Comment (bakoe) [reply CHANGELOG.md]: Absolutely! 👍 I created a corresponding separate issue (#15) for fixing this in the whole CHANGELOG.md file.

File: src/WebsiteInteractionSteps.php (hunk 10 above)
Comment (bakoe) [reply src/WebsiteInteractionSteps.php]: I did't find a way to achieve this behavior using `sprintf()` that eliminates the need for a ternary operator – thus, adding a `sprintf()` call would only increase the complexity with the options I tried. But thanks for the suggestion!

File: src/ConditionSteps.php (hunk 5 above)
Comment (bakoe) [reply src/ConditionSteps.php]: Yep, totally right! Fixed and now verified via a unit test. ✨

**Comment by bakoe** [comment]
@OskarStark I've integrated the feedback from your PR review (thanks for doing this!).
Please have another look, if you'd like. Thanks! 

/cc @Apfelschorle2017
